roles/populate_mirror_registry/tasks/prerequisites.yml

```diff
@@ -16,39 +29,73 @@
     - "{{ downloads_path }}"
     - "{{ downloads_path }}/{{ openshift_full_version }}"
 
+- name: Check if artifacts from release image have already been downloaded
+  stat:
+    path: "{{ downloads_path }}/{{ openshift_full_version }}/release.txt"
+  register: downloaded_artifacts
+
+# extract artifacts from the OCP release image instead of downloading them
+# from the release URL
+# This OCP release image provides the artifacts required (OCP clients, installer,
+# release.txt, among others)
+- name: Extract artifacts from release image
+  become: true
+  block:
+    # we need to temporarily download ocp stable binary to extract the artifacts
+    - name: Create tmp directory for oc stable client
+      tempfile:
+        state: directory
+        prefix: oc.
+      register: oc_tmp_dir
+
+    - name: Download stable client
+      unarchive:
+        src: "{{ stable_release_url }}/{{ stable_oc_tar }}"
+        dest: "{{ oc_tmp_dir.path }}"
+        mode: 0755
+        remote_src: yes
+
+    - name: Extract artifacts from release image (x86_64)
+      command:
+        cmd: >
+          "{{ oc_tmp_dir.path }}/oc" adm release extract
+            --registry-config "{{ config_file_path }}/{{ pull_secret_file_name }}"
+            --tools
+            --from "{{ release_image_item.remote | quote }}"
+            --to "{{ downloads_path }}/{{ openshift_full_version }}"
+      when: arm_release_image_key not in image_hashes
+
+    - name: Extract artifacts from release image (arm64)
```

Comment:
@ramperher Thanks for the explanation and the corrections. I'd think you would only want extract those artefacts if the bastion was arm? My reasoning behind this is that they are run on the bastion so the binaries should be for the bastion's architecture.

I think there are two options here: 

Either you can have var (something like `tool_extract_image`) which defaults to x86_64 release and document that if the bastion is arm they should change the release image and `stable_opm_tar`. 

Or check the ansible facts what the arch of the bastion is and use the correct release image and `stable_opm_tar` accordingly.

Reply:
Yes, that's it, I was trying to do something similar to what it is done [here](https://github.com/redhat-partner-solutions/crucible/blob/main/roles/populate_mirror_registry/tasks/populate_registry.yml#L53),  where the same check is done. But effectively, this is just to separate the cases of x86_64 and arm in two different tasks.

In our case, as we're always using x86_64, we were not paying attention to arm64. But now, I can see your point, and also that the `stable_oc/opm_tar` should point to arm one in case of being on arm64.

I'm going to follow your first approach, because in our case, we're not gathering Ansible facts by default, so this may affect to the performance of our playbooks. I'll rename the variable to `release_arch`, just to have a meaningful variable name, and I will refactor the `stable_oc/opm_tar` and the Extract artifacts tasks to simplify the logic a little bit more.

I'll also squash the commits after this, thanks!

Reply:
Now I've realized something: `oc_tar` changes depending on the architecture, you can see this in examples [(1)](https://mirror.openshift.com/pub/openshift-v4/clients/ocp/stable/) for a stable version and [(2)](https://openshift-release-artifacts.apps.ci.l2s4.p1.openshiftapps.com/4.13.0-0.nightly-2023-01-11-225449/) for a nightly version. But in the case of opm, where we always have to go to the stable repository (it is not available in the nightly repo), we don't have a distinction between x86_64 and arm64, so there, we'll directly download the only binary available.

Reply:
You can probably use `ansible_architecture`

Reply:
Ah I didn't realise that about OPM. Given that the tar has not arch in the name and without ensuring/testing that OPM works on an arm64 machine. I think the best route is to assume that the machine is `x86_64` as the previous code did. And remove the arm stuff entirely.


The issue I'm trying to avoid is it looking like it should work for arm and then blowing up because OPM doesn't work. It also simplifies the code which is nice.